Add job priorities to ActiveJob

wvengen opened the pull request:
The Active Job documentation [mentions](http://api.rubyonrails.org/classes/ActiveJob/QueueAdapters.html) that some backends allow setting priority per job. This PR adds support for setting job-level priorites to Active Job.
- Similar to queue_name, the priority can be set on the class level and can be overridden per job.
- This only brings support to job-based priorities, not queue-based priorities. (Though the same syntax could be used, with the backend checking that the same priority is always used per queue.)
- The default priority for delayed_job is 0, for que 100. To keep backward-compatibility, the job priority is `nil` by default, but this can be overridden by setting default_priority. [Both](https://github.com/collectiveidea/delayed_job/blob/ce88693429188a63793b16daaab67056a4e4e0bf/lib/delayed/backend/base.rb#L13) [backends](https://github.com/chanks/que/blob/a2a46fbfd9d336f5514434bdddc0b33bbc887fd6/lib/que/job.rb#L51) use the default priority when it is `nil`.
- Backends that don't support priorities will silently ignore any priority setting. Perhaps an exception could be thrown (like scheduled jobs with the inline backend), but as most backends will need to do that, I guess some refactoring is needed. Since this is my first contribution, I've kept it with this for now.

wvengen committed:
Add job priorities to ActiveJob

avokhmin commented:
:+1: , @wvengen can you do a rebase?

recurser commented:
:+1:

seuros commented:
Just looked at this more closely, the majority of the adapters don't support this feature :disappointed: .

wvengen commented:
@avokhmin yes, I'll have a look!

wvengen commented:
Rebased on +master+.
The only real change is that now priority `nil` is passed to the backend when none is given (which resulted in slightly simpler code). Both backends handle this as equal to not passing the priority argument (I've updated the PR description to reflect this).

abhchand commented:
:+1: Great addition

Looks like this needs another rebase. Any word on progress in getting this merged?

wvengen commented:
Rebased on `master`.

kaspth commented:
I'm :-1: on this given that the majority of the adapters don't support priorities (perhaps there's a reason for that :wink:).

wvengen commented:
@kaspth I'd say that there are different angles to handling job priority. You can use multiple queues with either multiple workers (most implementations), or workers that handle queues in a specific order (like sidekiq), or assign a priority to jobs (like que). In advanced cases you might even want to use a combination of those.

abhchand commented:
Thanks @wvengen!

deneuxa commented:
+1 for this feature please :)

wvengen commented:
Thanks for merging!

hale commented:
Thanks everyone. What's the milestone for this change?

kaspth commented:
Rails 5 :smile:

hale commented:
:+1:

jfine commented:
Awesome. Thanks @wvengen!

adeeb1 commented:
Sorry to comment on an old PR, but is this supported in or backported to Rails 4.2.7? I am using Delayed Job, and setting the priority in the following manner doesn't set it: `MyJob.set(priority: 2).perform_later`. The priority is still set to 0.

wvengen commented:
@adeeb1 @deneuxa Did you look at the ActiveJob [CHANGELOG](https://github.com/rails/rails/blob/v4.2.7.1/activejob/CHANGELOG.md) for 4.2.7? It doesn't list priorities. If you want them, the way to go is moving to Rails 5.

deneuxa commented:
thanks @wvengen we will try !
